File: samples/Linux/netcore_for_Linux_x64_only/netcore1.0/NonTableManagerSamples/Login/src/SessionStatusListener.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using fxcore2;

namespace Login
{
    class SessionStatusListener : IO2GSessionStatus
    {
        private string mSessionID;
        private string mPin;
        private bool mConnected;
        private bool mDisconnected;
        private bool mError;
        private O2GSession mSession;
        private EventWaitHandle mSyncSessionEvent;

        /// <summary>
        /// ctor
        /// </summary>
        /// <param name="session"></param>
        public SessionStatusListener(O2GSession session, string sSessionID, string sPin)
        {
            mSession = session;
            mSessionID = sSessionID;
            mPin = sPin;
            Reset();
            mSyncSessionEvent = new EventWaitHandle(false, EventResetMode.AutoReset);
        }

        public bool Connected
        {
            get
            {
                return mConnected;
            }
        }

        public bool Disconnected
        {
            get
            {
                return mDisconnected;
            }
        }

        public bool Error
        {
            get
            {
                return mError;
            }
        }

        public void Reset()
        {
            mConnected = false;
            mDisconnected = false;
            mError = false;
        }

        public bool WaitEvents()
        {
            return mSyncSessionEvent.WaitOne(30000);
        }

        public void onSessionStatusChanged(O2GSessionStatusCode status)
        {
            Console.WriteLine("Status: " + status.ToString());
            switch (status)
            {
                case O2GSessionStatusCode.TradingSessionRequested:
                    if (string.IsNullOrEmpty(mSessionID))
                    {
                        Console.WriteLine("Argument for trading session ID is missing");
                    }
                    else
                    {
                        mSession.setTradingSession(mSessionID, mPin);
                    }
                    break;
                case O2GSessionStatusCode.Connected:
                    mConnected = true;
                    mDisconnected = false;
                    mSyncSessionEvent.Set();
                    break;
                case O2GSessionStatusCode.Disconnected:
                    mConnected = false;
                    mDisconnected = true;
                    mSyncSessionEvent.Set();
                    break;
            }
        }

        public void onLoginFailed(string error)
        {
            Console.WriteLine("Login error: " + error);
            mError = true;
        }
    }
}

File: samples/macOS/java/NonTableManagerSamples/CreateOrderBySymbol/src/SessionStatusListener.java
package common;

import java.util.concurrent.Semaphore;
import java.util.concurrent.TimeUnit;

import com.fxcore2.*;

public class SessionStatusListener implements IO2GSessionStatus {
    private O2GSession mSession;
    private String mSessionID;
    private String mPin;
    private boolean mConnected;
    private boolean mDisconnected;
    private boolean mError;
    private Semaphore mSemaphore;

    // ctor
    public SessionStatusListener(O2GSession session, String sSessionID, String sPin) {
        mSession = session;
        mSessionID = sSessionID;
        mPin = sPin;
        reset();
        mSemaphore = new Semaphore(0);
    }

    public boolean isConnected() {
        return mConnected;
    }

    public boolean isDisconnected() {
        return mDisconnected;
    }

    public boolean hasError() {
        return mError;
    }

    public void reset() {
        mConnected = false;
        mDisconnected = false;
        mError = false;
    }

    public boolean waitEvents() throws Exception {
        return mSemaphore.tryAcquire(30, TimeUnit.SECONDS);
    }

    public void onSessionStatusChanged(O2GSessionStatusCode status) {
        System.out.println("Status: " + status.toString());
        switch (status) {
        case TRADING_SESSION_REQUESTED:
            if (mSessionID.isEmpty()) {
                System.out.println("Argument for trading session ID is missing");
            } else {
                mSession.setTradingSession(mSessionID, mPin);
            }
            break;
        case CONNECTED:
            mConnected = true;
            mDisconnected = false;
            mSemaphore.release();
            break;
        case DISCONNECTED:
            mConnected = false;
            mDisconnected = true;
            mSemaphore.release();
            break;
        }
    }

    public void onLoginFailed(String sError) {
        System.out.println("Login error: " + sError);
        mError = true;
    }
}
